CmdPal: Adding

michaeljolley opened the pull request:
Targets: #40382

- Adds support for "single-select" filters to DynamicListPage.
- Filters can contain icons
- Filter list can contain separators

Selecting a filter will update the `CurrentFilterIds` property of the ListPage's `Filters` property. Extension developers can opt-in to using filters by specifying a `Filters` property and responding to its changes.  An example can be found in the Windows Services built-in extension:

```C#
internal sealed partial class ServicesListPage : DynamicListPage
{
    public ServicesListPage()
    {
        Icon = Icons.ServicesIcon;
        Name = "Windows Services";

        var filters = new ServiceFilters();
        filters.PropChanged += Filters_PropChanged;
        Filters = filters;
    }

    private void Filters_PropChanged(object sender, IPropChangedEventArgs args) => RaiseItemsChanged();

    public override void UpdateSearchText(string oldSearch, string newSearch) => RaiseItemsChanged();

    public override IListItem[] GetItems()
    {
        var items = ServiceHelper.Search(SearchText, Filters.CurrentFilterIds).ToArray();

        return items;
    }
}

public partial class ServiceFilters : Filters
{
    public ServiceFilters()
    {
        CurrentFilterIds = ["all"];
    }

    public override IFilterItem[] GetFilters()
    {
        return [
            new Filter() { Id = "all", Name = "All Services" },
            new Separator(),
            new Filter() { Id = "running", Name = "Running", Icon = Icons.GreenCircleIcon },
            new Filter() { Id = "stopped", Name = "Stopped", Icon = Icons.RedCircleIcon },
            new Filter() { Id = "paused", Name = "Paused", Icon = Icons.PauseIcon },
        ];
    }
}
```

## Example of Windows Services extension in action

https://github.com/user-attachments/assets/3ffb3705-05d1-4614-9ba5-49b5ecf84dbf

davidegiacometti commented:
I love this and I would love a shortcut to cycle through filters: it would allow me to apply filters without use mouse (similar to CTRK+K for opening more commands) 🚀

michaeljolley commented:
My video showed me clicking, but it is the next tab location from the search bar. So tab, enter, up/down would get you there quick. I suppose we could add a shortcut to expand it as well. We'd have to determine what we want that to be. Maybe Ctrl+F?

michaeljolley commented:
> Hello, I just gave it a quick try, and it looks really good! I’d like to suggest one improvement: would you consider making the controls available as drop-downs (for example, open it with Alt+Down arrow or F4, or mouse wheel up/down over "closed" filter button)?

You mean beyond the Ctrl+F shortcut I just added?

michaeljolley commented:
> > > Hello, I just gave it a quick try, and it looks really good! I’d like to suggest one improvement: would you consider making the controls available as drop-downs (for example, open it with Alt+Down arrow or F4, or mouse wheel up/down over "closed" filter button)?
> > 
> > 
> > You mean beyond the Ctrl+F shortcut I just added?
> 
> Yes. When I was trying it, I instinctively pressed Alt + Down, since it looks like a dropdown—I expected it to behave like one.
> 
> If you'd like, I can handle the necessary work to add this, so you don’t have to spend time on it. 😇

This got me thinking. "Why aren't I just using a ComboBox here, rather than copying the stuff from the ContextMenu?"

Refactored, this to be a ComboBox so Alt+DownArrow and F4 all work to expand it. Users can also use up/down arrow keys to flip through them.

jiripolasek commented:
> > > > Hello, I just gave it a quick try, and it looks really good! I’d like to suggest one improvement: would you consider making the controls available as drop-downs (for example, open it with Alt+Down arrow or F4, or mouse wheel up/down over "closed" filter button)?
> > > 
> > > 
> > > You mean beyond the Ctrl+F shortcut I just added?
> > 
> > 
> > Yes. When I was trying it, I instinctively pressed Alt + Down, since it looks like a dropdown—I expected it to behave like one.
> > If you'd like, I can handle the necessary work to add this, so you don’t have to spend time on it. 😇
> 
> This got me thinking. "Why aren't I just using a ComboBox here, rather than copying the stuff from the ContextMenu?"
> 
> Refactored, this to be a ComboBox so Alt+DownArrow and F4 all work to expand it. Users can also use up/down arrow keys to flip through them.

Works great! I can turn off my brain and let muscle memory drive it, and it works (tab, down, down, untab, write query).

Did you shelve the idea of multi-select filtering? I actually have some use cases where I’d find that useful.

htcfreek commented:
Filter might be interesting for windows search ext. Especially if we can add the the filter to the query. I have the following in my mind: All, Directories (type:directory), Files (type:file).

michaeljolley commented:
I did shelve the idea of multi-select for now. Happy to bring it back but that would rule out ComboBox.

jiripolasek commented:
> I did shelve the idea of multi-select for now. Happy to bring it back but that would rule out ComboBox.

I think there’s some utility here—for example, a Show hidden files toggle in the file browser. I’ve also been thinking about this from another angle, specifically regarding page- or extension-wide commands, and there’s definitely some overlap. I’ll create a feature issue and jot down my thoughts there to keep this discussion focused.

zadjii-msft reviewed (changes requested):
okay I didn't review too closely, but I'm worried that the initialization of these properties on weird threads is why you were seeing those COM issues - most of the COM objects are created off of the UI thread, so they can't then be accessed on the UI thread (its a whole thing).

This probably doesn't matter for in-proc extensions, as they don't use OOP MBM to get the objects
